File: go/test/python_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/cloudsmith-sdk/go"
	"github.com/voxgig-sdk/cloudsmith-sdk/go/core"

	vs "github.com/voxgig-sdk/cloudsmith-sdk/go/utility/struct"
)

func TestPythonEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Python(nil)
		if ent == nil {
			t.Fatal("expected non-nil PythonEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := pythonBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "python." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_PYTHON_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		pythonRef01Ent := client.Python(nil)
		pythonRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "python"}, setup.data), "python_ref01"))
		pythonRef01Data["identifier"] = setup.idmap["identifier01"]
		pythonRef01Data["owner"] = setup.idmap["owner01"]

		pythonRef01DataResult, err := pythonRef01Ent.Create(pythonRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		pythonRef01Data = core.ToMapAny(pythonRef01DataResult)
		if pythonRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		pythonRef01Match := map[string]any{
			"identifier": setup.idmap["identifier01"],
			"owner": setup.idmap["owner01"],
		}

		pythonRef01ListResult, err := pythonRef01Ent.List(pythonRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		pythonRef01List, pythonRef01ListOk := pythonRef01ListResult.([]any)
		if !pythonRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", pythonRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(pythonRef01List), map[string]any{"id": pythonRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		pythonRef01DataUp0Up := map[string]any{
			"identifier": setup.idmap["identifier"],
			"owner": setup.idmap["owner"],
		}

		pythonRef01MarkdefUp0Name := "auth_mode"
		pythonRef01MarkdefUp0Value := fmt.Sprintf("Mark01-python_ref01_%d", setup.now)
		pythonRef01DataUp0Up[pythonRef01MarkdefUp0Name] = pythonRef01MarkdefUp0Value

		pythonRef01ResdataUp0Result, err := pythonRef01Ent.Update(pythonRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		pythonRef01ResdataUp0 := core.ToMapAny(pythonRef01ResdataUp0Result)
		if pythonRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if pythonRef01ResdataUp0[pythonRef01MarkdefUp0Name] != pythonRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", pythonRef01MarkdefUp0Name, pythonRef01ResdataUp0[pythonRef01MarkdefUp0Name])
		}

		// LOAD
		pythonRef01MatchDt0 := map[string]any{}
		pythonRef01DataDt0Loaded, err := pythonRef01Ent.Load(pythonRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if pythonRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func pythonBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "python", "PythonTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read python test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse python test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"python01", "python02", "python03", "package01", "package02", "package03", "repo01", "repo02", "repo03", "identifier01", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CLOUDSMITH_TEST_PYTHON_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CLOUDSMITH_TEST_PYTHON_ENTID": idmap,
		"CLOUDSMITH_TEST_LIVE":      "FALSE",
		"CLOUDSMITH_TEST_EXPLAIN":   "FALSE",
		"CLOUDSMITH_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CLOUDSMITH_TEST_PYTHON_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add identifier alias for update test.
	if idmapResolved["identifier"] == nil {
		idmapResolved["identifier"] = idmapResolved["identifier01"]
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}

	if env["CLOUDSMITH_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CLOUDSMITH_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCloudsmithSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/PythonEntityTest.php
<?php
declare(strict_types=1);

// Python entity test

require_once __DIR__ . '/../cloudsmith_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class PythonEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = CloudsmithSDK::test(null, null);
        $ent = $testsdk->Python(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "python" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = CloudsmithSDK::test($seed, null);
        $seen = iterator_to_array($base->Python(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = CloudsmithConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = CloudsmithSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Python(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = python_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "python." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_PYTHON_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $python_ref01_ent = $client->Python(null);
        $python_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.python"), "python_ref01"));
        $python_ref01_data["identifier"] = $setup["idmap"]["identifier01"];
        $python_ref01_data["owner"] = $setup["idmap"]["owner01"];

        $python_ref01_data_result = $python_ref01_ent->create($python_ref01_data, null);
        $python_ref01_data = Helpers::to_map($python_ref01_data_result);
        $this->assertNotNull($python_ref01_data);

        // LIST
        $python_ref01_match = [
            "identifier" => $setup["idmap"]["identifier01"],
            "owner" => $setup["idmap"]["owner01"],
        ];

        $python_ref01_list_result = $python_ref01_ent->list($python_ref01_match, null);
        $this->assertIsArray($python_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($python_ref01_list_result),
            ["id" => $python_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $python_ref01_data_up0_up = [
            "identifier" => $setup["idmap"]["identifier"],
            "owner" => $setup["idmap"]["owner"],
        ];

        $python_ref01_markdef_up0_name = "auth_mode";
        $python_ref01_markdef_up0_value = "Mark01-python_ref01_" . $setup["now"];
        $python_ref01_data_up0_up[$python_ref01_markdef_up0_name] = $python_ref01_markdef_up0_value;

        $python_ref01_resdata_up0_result = $python_ref01_ent->update($python_ref01_data_up0_up, null);
        $python_ref01_resdata_up0 = Helpers::to_map($python_ref01_resdata_up0_result);
        $this->assertNotNull($python_ref01_resdata_up0);
        $this->assertEquals($python_ref01_resdata_up0[$python_ref01_markdef_up0_name], $python_ref01_markdef_up0_value);

        // LOAD
        $python_ref01_match_dt0 = [];
        $python_ref01_data_dt0_loaded = $python_ref01_ent->load($python_ref01_match_dt0, null);
        $this->assertNotNull($python_ref01_data_dt0_loaded);

    }
}

function python_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/python/PythonTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = CloudsmithSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["python01", "python02", "python03", "package01", "package02", "package03", "repo01", "repo02", "repo03", "identifier01", "owner01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("CLOUDSMITH_TEST_PYTHON_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "CLOUDSMITH_TEST_PYTHON_ENTID" => $idmap,
        "CLOUDSMITH_TEST_LIVE" => "FALSE",
        "CLOUDSMITH_TEST_EXPLAIN" => "FALSE",
        "CLOUDSMITH_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["CLOUDSMITH_TEST_PYTHON_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["identifier"])) {
        $idmap_resolved["identifier"] = $idmap_resolved["identifier01"];
    }
    if (!isset($idmap_resolved["owner"])) {
        $idmap_resolved["owner"] = $idmap_resolved["owner01"];
    }

    if ($env["CLOUDSMITH_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["CLOUDSMITH_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new CloudsmithSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["CLOUDSMITH_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["CLOUDSMITH_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
